proposals/0016-c-interface-compiler-library.md

```diff
@@ -50,7 +50,495 @@ older DXC-based solutions to adopt clang as the preferred HLSL compiler.
+Code snippets on how to compile a shader using the COM library are located
+in [DXC api code snippets](#dxc-code-snippets).
+
+
+The following interfaces are used to work with data being passed to/from the
+library.
+
+#### Buffers
+```c++
+// Structure for supplying bytes or text input to Dxc APIs. Represents both
+// text (DxcText) and non-text byte buffers (DxcBuffer).
+struct DxcBuffer {
+  LPCVOID Ptr;
+  SIZE_T Size;
+  UINT Encoding;
+};
+typedef DxcBuffer DxcText;
+
+// General purpose buffers
+struct IDxcBlob : public IUnknown {
+public:
+  LPVOID GetBufferPointer();
+  SIZE_T GetBufferSize();
+};
+
+// String buffers that guarantee null-terminated text and the stated encoding
+struct IDxcBlobEncoding : public IDxcBlob {
+public:
+  HRESULT GetEncoding(BOOL *pKnown, UINT32 *pCodePage);
+};
+
+struct IDxcBlobUtf16 : public IDxcBlobEncoding {
+public:
+  LPCWSTR GetStringPointer();
+  SIZE_T GetStringLength();
+};
+
+struct IDxcBlobUtf8 : public IDxcBlobEncoding {
+public:
+  LPCSTR GetStringPointer();
+  SIZE_T GetStringLength();
+};
+```
+#### Status and other data
+Errors, status, and other kinds of data are returned as an IDxcResult
+interface. A unique DXC_OUT_KIND is defined for each different type of data
+being returned.  The type of data is determined by the interface method used to
+produce the IDxcResult.
+
+Example:
+IDxcCompiler3::Compile() and IDxcCompiler3::Disassemble() both return an 
+IDxcResult.
+
+```c++
+typedef enum DXC_OUT_KIND {
+  DXC_OUT_NONE = 0,
+  DXC_OUT_OBJECT = 1,         // IDxcBlob - Shader or library object
+  DXC_OUT_ERRORS = 2,         // IDxcBlobUtf8 or IDxcBlobUtf16
+  DXC_OUT_PDB = 3,            // IDxcBlob
+  DXC_OUT_SHADER_HASH = 4,    // IDxcBlob - DxcShaderHash of shader or shader
+                              // with source info (-Zsb/-Zss)
+  DXC_OUT_DISASSEMBLY = 5,    // IDxcBlobUtf8/16 - from Disassemble
+  DXC_OUT_HLSL = 6,           // IDxcBlobUtf8/16 - from Preprocessor or Rewriter
+  DXC_OUT_TEXT = 7,           // IDxcBlobUtf8/16 - other text, such as -ast-dump
+                              // or -Odump
+  DXC_OUT_REFLECTION = 8,     // IDxcBlob - RDAT part with reflection data
+  DXC_OUT_ROOT_SIGNATURE = 9, // IDxcBlob - Serialized root signature output
+  DXC_OUT_EXTRA_OUTPUTS  = 10,// IDxcExtraResults - Extra outputs
+
+  DXC_OUT_FORCE_DWORD = 0xFFFFFFFF
+} DXC_OUT_KIND;
+
+struct IDxcOperationResult : public IUnknown {
+  HRESULT GetStatus(HRESULT *pStatus);
+
+struct IDxcResult : public IDxcOperationResult {
+  BOOL HasOutput(DXC_OUT_KIND dxcOutKind);
+  HRESULT GetOutput(DXC_OUT_KIND dxcOutKind,
+                    REFIID iid, void **ppvObject,
+                    IDxcBlobUtf16 **ppOutputName);
+  UINT32 GetNumOutputs();
+  DXC_OUT_KIND GetOutputByIndex(UINT32 Index);
+  DXC_OUT_KIND PrimaryOutput();
+};
+```
+#### Helper library
+The shader compiler library also provides a helper library IDxcUtils that is
```

Comment:
There's a lib called IDxcUtils?  I'm not sure that's the case is it?  How do I get an IDxcUtils pointer?

Reply:
Looks like IDxcUtils fulfils a factory-like role in this API.

Reply:
I showed how that can be obtained in the snippets.  CreateInstance is called to get that object